specs/WebMessageObjects.md

```diff
@@ -0,0 +1,355 @@
+
+```html
+<!-- File upload location -->
+<input type="file" id="files" multiple />
+```
+
+```javascript
+const input = document.getElementById('files');
+input.addEventListener('change', function() {
+    const currentFiles = input.files;
+    chrome.webview.postMessageWithAdditionalObjects("FilesDropped", currentFiles);
+});
+```
+
+```cpp
+m_webView->add_WebMessageReceived(
+    Callback<ICoreWebView2WebMessageReceivedEventHandler>(
+        [this](ICoreWebView2* sender, ICoreWebView2WebMessageReceivedEventArgs* args)
+        {
+            wil::com_ptr<ICoreWebView2WebMessageReceivedEventArgs2> args2 =
+                wil::com_ptr<ICoreWebView2WebMessageReceivedEventArgs>(args)
+                    .query<ICoreWebView2WebMessageReceivedEventArgs2>();
+            wil::com_ptr<ICoreWebView2WebMessageObjectCollectionView>
+                objectsCollection;
+            args2->get_AdditionalObjects(&objectsCollection);
+            unsigned int length;
+            objectsCollection->get_Count(&length);
+            std::vector<std::wstring> paths;
+
+            for (unsigned int i = 0; i < length; i++)
+            {
+                wil::com_ptr<IUnknown> object;
+                objectsCollection->GetValueAtIndex(i, &object);
+
+                wil::com_ptr<ICoreWebView2File> file =
+                    object.query<ICoreWebView2File>();
+                if (file)
+                {
+                    // Add the file to message to be sent back to webview
+                    wil::unique_cotaskmem_string path;
+                    file->get_Path(&path);
+                    paths.push_back(path.get());
+                }
+            }
+
+            ProcessPaths(paths);
+            return S_OK;
+        })
+        .Get(),
+    &m_webMessageReceivedToken);
+```
+
+```c#
+webView.CoreWebView2.WebMessageReceived += WebView_WebMessageReceivedHandler;
+void WebView_WebMessageReceivedHandler(object sender, CoreWebView2WebMessageReceivedEventArgs args)
+{
+    List<string> paths = new List<string>();
+    foreach (var object in args.AdditionalObjects)
+    {
+        if (object is CoreWebView2File)
+        {
+            paths.Add(((CoreWebView2File)object).Path);
+        }
+    }
+    ProcessPaths(paths);
+}
+```
+
+# API Details
+## SDK API
+```c#
+/// Representation of a DOM
+/// (File)[https://developer.mozilla.org/en-US/docs/Web/API/File] object
+/// passed via WebMessage. You can use this object to obtain the path of a
+/// File dropped on WebView2.
+/// \snippet ScenarioDragDrop.cpp DroppedFilePath
+[uuid(f2c19559-6bc1-4583-a757-90021be9afec), object, pointer_default(unique)]
+interface ICoreWebView2File : IUnknown {
+  /// Get the absolute file path.
+  [propget] HRESULT Path([out, retval] LPWSTR* path);
+}
+
+/// Read-only collection of generic objects
+[uuid(b547d2d4-b8d5-4278-a544-c54e66b5f45d), object, pointer_default(unique)]
+interface ICoreWebView2ObjectCollectionView : IUnknown {

```

Comment:
Does WebView2 have a guideline for using "View" to indicate "you can't change this set"?   I see [ICoreWebView2StringCollection](https://learn.microsoft.com/en-us/microsoft-edge/webview2/reference/win32/icorewebview2stringcollection?view=webview2-1.0.1418.22) and [ICoreWebView2ClientCertificateCollection](https://learn.microsoft.com/en-us/microsoft-edge/webview2/reference/win32/icorewebview2clientcertificatecollection?view=webview2-1.0.1418.22) don't have Insert/Remove, but still don't have the View suffix.    I didn't find any results for GetValueAtIndex with a class using the View suffix.

Reply:
We do use "View" to denote some read-only objects (i.e. [ICoreWebView2ResponseView](https://learn.microsoft.com/en-us/microsoft-edge/webview2/reference/win32/icorewebview2webresourceresponseview?view=webview2-1.0.1418.22) , but as you pointed out we haven't used it on the Collection type (we probably should have). This would establish a new precedent for Collection type if we go this route.

Reply:
New rule for COM interfaces: Read-only collections are `*CollectionView` and mutable collections are `*Collection`.

Looks good for this.